### PR TITLE
Implemented Filtering by Country

### LINKED ISSUE #5257: Filter by Location (country)
From support mail:

Hello guys. Please add filter by Location (country). Could be very usefull to sort stored caches to different lists. Thanks.

### PR DESCRIPTION
**Pull request by jakedot**
Unconstrained, over all Countries extracted from the Location field on all caches.

Fixes #5257 
See #8210 for history.

### CODE REVIEW

**Comment by moving-bits** [comment]
ok to test

**Comment by Lineflyer** [comment]
ok to test


@moving-bits @bekuno
We need to look up why you are not allowed to talk to the CI.

File: main/src/cgeo/geocaching/filter/CountryFilterFactory.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+    @Override
+    @NonNull
+    public List<IFilter> getFilters() {
+
+        final Map<String, IFilter> filters = new HashMap<>();
+
+        final String separator = ", ";
+
+        // TODO: use DataStore.getStoredLocationsForSet(Set<String>)
```
Comment (jakedot) [inline main/src/cgeo/geocaching/filter/CountryFilterFactory.java:68]: Here, I need to somehow get the list of current geocodes.

File: main/src/cgeo/geocaching/storage/DataStore.java [hunk 2]
```diff
@@ -3408,6 +3411,86 @@ public static void saveLists(final Collection<Geocache> caches, final Set<Intege
...
+        } catch (final SQLiteDoneException ignored) {
+            // Do nothing, it only means we have no information on the cache
+        } catch (final Exception e) {
+            Log.e("DataStore.getAllStoredLocations", e);
+        }
+
+        return locations;
+    }
+
+    public static Set<String> getStoredLocationsForSet(Set<String> geocodes) {
```
Comment (jakedot) [inline main/src/cgeo/geocaching/storage/DataStore.java:3460]: this should be able to utilize the DB for getting all locations for a set of geocodes, this way we can skip creating Geocache objects

File: main/src/cgeo/geocaching/storage/DataStore.java [hunk 3]
```diff
@@ -3408,6 +3411,86 @@ public static void saveLists(final Collection<Geocache> caches, final Set<Intege
         }
     }
 
+    public static Set<String> getStoredLocationsInList(final int listId) {
+
```
Comment (jakedot) [inline main/src/cgeo/geocaching/storage/DataStore.java:3415]: this could get all locations in a stored list I guess. may be dropped if wholly unnecessary.

**Comment by moving-bits** [comment]
@jakedot 
Please follow the "Details" link of the failed "default - Build finished." line to see the remaining CheckStyle issues (or other things that lead to a broken build).

File: main/src/cgeo/geocaching/filter/CountryFilterFactory.java (hunk 1 above)
Comment (jakedot) [reply main/src/cgeo/geocaching/filter/CountryFilterFactory.java:68]: @moving-bits any idea?

Comment (moving-bits) [reply main/src/cgeo/geocaching/filter/CountryFilterFactory.java:68]: @jakedot  Nothings springs to my mind immediately, but I haven't looked into the source code yet. Maybe have a look at the position where the filter gets instantiated - can you add the set as a new parameter there?

**Comment by jakedot** [comment]
It should be possible, but I still have no clue where the set can be
retrieved from. The filter is created in the FilterActivity, which doesn't
seem to have it. I guess we would need to pass it in with an Intent?

**Comment by moving-bits** [comment]
So filters are static? Then we need to inject the current set later, when the list is created. (if possible)

**Pull request merged**